core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java

```diff
@@ -17,15 +116,1084 @@
+         } else {
+            Object value = entry.getValue();
+            if (trace) {
+               log.tracef("Found %s -> %s", toStr(key), toStr(value));
+            }
+            map.put(key, value);
+         }
+      }
+      return map;
+   }
+
+   private static class BackingKeySet<K, V> extends AbstractCollection<K> implements CacheSet<K> {
+      private final boolean isRemoteIteration;
+      private final Cache<K, V> cache;
+      private final InternalDataContainer<K, V> dataContainer;
+      private final KeyPartitioner keyPartitioner;
+
+      BackingKeySet(Cache<K, V> cache, InternalDataContainer<K, V> dataContainer, KeyPartitioner keyPartitioner, boolean isRemoteIteration) {
+         this.cache = cache;
+         this.dataContainer = dataContainer;
+         this.keyPartitioner = keyPartitioner;
+         this.isRemoteIteration = isRemoteIteration;
+      }
+
+      @Override
+      public CloseableIterator<K> iterator() {
+         if (isRemoteIteration) {
+            // Don't add the extra wrapping for removal
+            return new IteratorMapper<>(dataContainer.iterator(), Map.Entry::getKey);
+         }
+         return new IteratorMapper<>(new DataContainerRemoveIterator<>(cache, dataContainer), Map.Entry::getKey);
+      }
+
+      @Override
+      public CloseableSpliterator<K> spliterator() {
+         // Spliterator doesn't support remove so just return it without wrapping
+         return new SpliteratorMapper<>(dataContainer.spliterator(), Map.Entry::getKey);
+      }
+
+      @Override
+      public int size() {
+         return dataContainer.size();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         return dataContainer.containsKey(o);
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         return cache.remove(o) != null;
+      }
+
+      private CacheStream<K> doStream(boolean parallel) {
+         return new LocalCacheStream<>(new SegmentedKeyStreamSupplier<>(cache, keyPartitioner, dataContainer), parallel,
+               cache.getAdvancedCache().getComponentRegistry());
+      }
+
+      @Override
+      public CacheStream<K> stream() {
+         return doStream(false);
+      }
+
+      @Override
+      public CacheStream<K> parallelStream() {
+         return doStream(true);
+      }
+   }
+
+   @Override
+   public Object visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
+      long flagBitSet = command.getFlagsBitSet();
+      // We have to check for the flags when we do perform - as interceptor could change this while going down
+      // the stack
+      boolean isRemoteIteration = EnumUtil.containsAny(flagBitSet, FlagBitSets.REMOTE_ITERATION);
+      return new BackingKeySet<>(cacheWithFlags(flagBitSet), dataContainer, keyPartitioner, isRemoteIteration);
+   }
+
+   static class BackingEntrySet<K, V> extends AbstractCollection<CacheEntry<K, V>> implements CacheSet<CacheEntry<K, V>> {
```

Comment:
Please move the nested classes after all the methods

Reply:
Sure

Reply:
I was trying to keep them near where they were used, but I can move them all down.